File: App/FantasyGolf/src/main/java/League.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.FileWriter;
import java.io.IOException;

public class League {
    private ArrayList<Entry> teams = new ArrayList<Entry>();
    public String worstPlayer;
    private ArrayList<String> listOfPlayers;

    public League() {
        listOfPlayers = new ArrayList<>();
    }

    public void addEntry(Entry e) {
        teams.add(e);
    }

    public ArrayList<Entry> getEntries() {
        return this.teams;

    }

    public void score() {

        for (Entry e : teams) {
            for (Player p : e.getEntry()) {
                listOfPlayers.add(p.getName());
            }
        }

        // find lowest ranked player to make top 25
        Results r = new Results();
        r.inputResultsAndRankings();
        LinkedList<Results.Player> top25 = r.top25ByRanking();
        String worst = "n/a";
        for (int i = top25.size() - 1; i >= 0; i--) {
            if (listOfPlayers.contains(top25.get(i).name)) {
                worst = top25.get(i).name;
            }
        }
        this.worstPlayer = worst;

        for (Entry e : teams) {
            e.score(worst);
        }

        // Sort the teams based on points scored
        Collections.sort(teams, new Comparator<Entry>() {
            public int compare(Entry a, Entry b) {
                return (Float.valueOf(b.getScore()).compareTo(Float.valueOf(a.getScore())));
            }
        });

        // Print results
        for (Entry e : teams) {
            System.out.println(e.getName() + ": " + e.getScore());

        }
    }

    /**
     * Takes the results of the tournament out creates a json file with all the results
     * @return String specifying whether the operation was done succesfully
     */
    public String makeJSONObject() {
        JSONObject jsonObject = new JSONObject();
        
        // Get the current date and time using Java 8 Date-Time API
        LocalDateTime currentDateTime = LocalDateTime.now();
        
        // Define the desired date and time format
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("EEEE, MMMM d h:mm a");

        // Format the current date and time using the formatter
        String formattedDateTime = currentDateTime.format(formatter);
        jsonObject.put("Date", formattedDateTime);

        // Create an array of JSON objects that will be the teams
        JSONArray entries = new JSONArray();
        
        try {
            //Go through each team's roster
            int place = 1;
            for (Entry team : this.getEntries()) {
                JSONObject entry = new JSONObject();
                entry.put("Name", team.getName());
                entry.put("Total Score", team.getScore());
                entry.put("Place", place++);
                JSONArray roster = new JSONArray();

                //Goes through each player and adds them to array of players
                for (Player p : team.getEntry()) {
                    JSONObject player = new JSONObject();
                    player.put("Name", p.getName());
                    player.put("Points scored", p.getPoints());
                    player.put("Finish", p.getFinish());
                    roster.put(player);
                }
                entry.put("Roster",roster);
                entry.put("AllCut", team.ALLCUT);
                entry.put("Worst25", team.WORST_IN_25);
                entries.put(entry);   
            }
            jsonObject.put("Teams",entries);
            
            
            
        } catch (JSONException e) {
            e.printStackTrace();
            return "Failed to make JSON Object"; 
        }

        // Specify the file path
        String filePath = "App/Database/Results.json";

        try (FileWriter fileWriter = new FileWriter(filePath)) {
            // Write the JSON object to the file
            fileWriter.write(jsonObject.toString());
        } catch (IOException e) {
            e.printStackTrace();
            return "Failed to make JSON file";
        }
    return "JSON file created successfully!";
    }
}

File: App/FantasyGolf/main/league.py
import json
from datetime import datetime
from zoneinfo import ZoneInfo

from teams import Team
import config
class League:

    # Constructor to initialize the object
    def __init__(self):
        self.teams = [] 
        self.worst_player = None    # Lowest ranked selected player to make top 25
        self.list_of_players = []
        self.player_scores = {}  # Dictionary to map player names to their scores
    
    def add_team(self, team):
        self.teams.append(team)
    
    def list_teams(self):
        for team in self.teams:
            print(team.name)

    def calculate_score(self):
        for team in self.teams:
            team.calculate_score(self.worst_player)
    
    def sort_teams(self):
        self.teams = sorted(self.teams, lambda team: team.get_score)
    
    def set_worst_player_in_top_25(self, worst):
        self.worst_player = worst
    
    def JSONify(self):
        '''
        Creates the JSON object with the results of the league to be sent to database
        '''
        # Sort teams by score 
        self.teams = sorted(self.teams, key=lambda team:team.get_score(), reverse=True)
        
        data = {}
        data["title"] = "leaderboard"
        
        now = datetime.now(ZoneInfo("America/New_York"))
        formatted_date = now.strftime("%A, %B %d %-I:%M %p")
        data["Date"] = formatted_date
        print(formatted_date)
        data["CURRENT_ROUND"] = config.CURRENT_ROUND

        data['worstTop25'] = self.worst_player
        
        teams_objects = []
        prev_score = -1
        place = 0
        numTies = 0
        for team in self.teams:
            team_obj = {}
            team_obj["Name"] = team.get_name()
            team_obj["Total Score"]  = team.get_score()
            team_obj["WorstRankedBonus"] = team.get_worst_ranked()
            if team.get_score() != prev_score:
                place += (1 + numTies)
                numTies = 0
            else:
                numTies += 1
            team_obj["Place"] = place
            prev_score = team.get_score()
            team_obj["AllCut"] = team.get_all_cut()
            teams_objects.append(team_obj)
            roster = []
            for player in team.get_roster():
                roster.append(player.JSONify())
            team_obj["Roster"] = roster
        
        data["Teams"] = teams_objects
        
        with open("resources/Results.json", "w") as f:
            json.dump(data, f, indent=4)
    
    def make_overall_leaderboard(self):
        '''
        If this is the masters, the overall leaderboard needs to be created
        '''
        data = {}
        data["title"] = "overall"
        teams_objects = []
        prev_score = -1
        numTies = 0
        place = 0
        for team in self.teams:
            team_obj = {}
            team_obj["name"] = team.get_name()
            team_obj["email"] = team.get_email()
            masters = team.get_score()
            team_obj["masters"] = masters
            team_obj["US"] = 0
            team_obj["open"] = 0 
            team_obj["pga"] = 0
            team_obj["total"] = masters
            if masters != prev_score:
                place += (1 + numTies)
                numTies = 0
            else:
                numTies += 1
            team_obj["place"] = place
            teams_objects.append(team_obj)
            prev_score = masters
            
        data["Teams"] = teams_objects
        with open("resources/Overall.json", "w") as f:
            json.dump(data, f, indent=4)
    
    def update_overall(self, overall):
        '''
        Updates the overall leaderboard with current scores 
        '''
        #add the score from the current tournament to the overall leaderboard
        for team in overall["Teams"]:
            matching_team = None
            for comp_team in self.teams:
                if team.get("email") == comp_team.get_email():
                    matching_team = comp_team
                    break
            if matching_team == None:
                print(f"no matching team for email: {team.get('email')}")
                return
            total = 0
            if config.current_tournament == config.TOURNAMENT.PGA:
                team['pga'] = matching_team.get_score()
            total += team['pga']
            
            if config.current_tournament == config.TOURNAMENT.OPEN_CHAMPIONSHIP:
                team['open'] = matching_team.get_score()
            total += team['open']
            
            if config.current_tournament == config.TOURNAMENT.US_OPEN:
                team['US'] = matching_team.get_score()
            total += team['US']
            
            if config.current_tournament == config.TOURNAMENT.OPEN_CHAMPIONSHIP:
                team['master'] = matching_team.get_score()  
            total += team['masters']
            
            team["total"] = total


        #sort the team objects by total and calculate places 
        overall["Teams"].sort(key=lambda team: team["total"], reverse=True)
        prev_total = -1
        numTies = 0
        place = 0
        for team in overall["Teams"]:
            if team['total'] != prev_total:
                place += (1 + numTies)
                numTies = 0
            else:
                numTies += 1
            prev_total = team['total']
            team["place"] = place
        
        with open("resources/Overall.json", "w") as f:
                json.dump(overall, f, indent=4)
